### DISCUSSION TITLE
Add WiFi control to any EVSE: Using HAA without the Arduino part?

### DISCUSSION THREAD

**Discussion by tokamac** [RavenSystem/esp-homekit-devices#2182] [General]
### Purpose of this open-source hack
**Enabling control access over Wi-Fi of any basic EVSE** (electric vehicle supply equipment, commonly called an "EV charger" or "wallbox") that has no connected smart feature in the first place.

### How
In all EV charging cables, there is a thin wire called the "Control Pilot" (CP). This wire carries a 1kHz square form signal at 12V constant voltage.

If the control pilot signal is absent (i.e. drops to 0V) then charging the EV is impossible (not acknowledged by the wallbox). Therefore, being able to interrupt the CP voltage at a distance over Wi-Fi allows access control (preventing unauthorised people to charge their EV). If integrated with a smart environment like HomeKit, the owner can also schedule the charging to take place at off-peak times using Automations. He can also properly interrupt charging at a distance, at any time and from anywhere.

This is the correct way to interrupt an EV's charge. Sure, alternately a Wi-Fi DIN circuit breaker in the electric panel can de-energize the entire EVSE and cut off the charge, but this is akin to a sudden power failure that neither the charging station nor the vehicle's electronics and software appreciate. Dropping the CP signal to 0V instead is a proper request to the wallbox to stop charging.

### Enhanced solution
The tutorial [https://ev-olution.yolasite.com/Sonoff-EVSE.php](https://ev-olution.yolasite.com/Sonoff-EVSE.php) shows how the output terminal of a Sonoff SV in isolated mode can be controlled over Wi-Fi as a potential-free (dry contact) switch to achieve such a goal. 
Moreover:
An Arduino Pro Mini is used to monitor the control pilot signal and Sonoff relay to disable charging by opening the relay again if the T2 plug is not connected to the EV within 20 seconds after activation of the relay, or 1 second after the plug has been disconnected from the vehicle. This ensures that the ability to charge an EV will not remain active if the user inadvertently activates the switch or forgets to switch it off after charging is complete. It also prevents unauthorized use of the charging station without the owner's knowledge or consent.

![sonoff 1](https://github.com/RavenSystem/esp-homekit-devices/assets/3685009/b21554f8-b512-4d32-a417-272d29d166ba)
![sonoff side 2](https://github.com/RavenSystem/esp-homekit-devices/assets/3685009/5d433daa-a14c-418c-8cec-30a914291f2d)
![sonoff connected](https://github.com/RavenSystem/esp-homekit-devices/assets/3685009/a5f29da1-c964-4125-ad80-9d74d2612c46)
Arduino pins: 
**2** - relay deactivation signal
**7** - EVSE control pilot signal state detection
**8** - relay state detection

Arduino code:

```
int status = 0; // connection to EV status;
int i = 0;
int n = 0;
int k = 0;
const long interval = 500;

void setup()
{
  Serial.begin(9600);
  pinMode(8, INPUT); // relay state detection; LOW-relay on, HIGH-relay off;
  pinMode(2, OUTPUT); // relay deactivation impulse (HIGH);
  pinMode(7, INPUT); // EVSE pilot state detection;
  digitalWrite(2, LOW);
}

void loop()
{
  if (digitalRead(8) == HIGH) {
    delay(200);
    status = 0;
  }
  k = 0;
  while ((digitalRead(8) == LOW && digitalRead(7) == HIGH) && status == 0) {
    for (i = 1; i <= 40; i += 1) {
      scan();
      if (k == 1){
        break;
      }
      if (i == 40) {
        digitalWrite(2, HIGH);
        delay(500);
        digitalWrite(2, LOW);
        delay(1000);
      }
    }
  }
  if (digitalRead(8) == LOW) {
    delay(200);
    status = 1;
  }
  k = 0;
  while ((digitalRead(8) == LOW && digitalRead(7) == HIGH) && status == 1) {
    for (n = 1; n <= 2; n += 1) {
      scan();
      if (k == 1){
        break;
      }
      if (n == 2) {
        digitalWrite(2, HIGH);
        delay(500);
        digitalWrite(2, LOW);
        delay(1000);
        status = 0;
      }
    }
  }
}
void scan(){
  unsigned long currentMillis = millis();

  while (millis() - currentMillis < interval) {
    if (digitalRead(7) == LOW || digitalRead(8) == HIGH){
      k = 1;
      return;
    }
  }
}
```

### Question
**Can the logic implemented in the Arduino in the example above be coded directly in HAA?** So that only the Sonoff SV is mandatory, and no Arduino is needed.

**Comment by tonysprenk** [comment]
I think it should be possible with HAA to change the state of the CP wire and read the state of the relays. I am not good at scripting for HAA so not sure what that would look like but leaving a comment here as I am also interested in this solution.

**Comment by tokamac** [comment]
HAA knows the state of the relay (open/close) already at any time. As the CP circuit is directly controlled by the relay, i.e. opened or closed, both status are equal. The trick here would be for HAA to know if the T2 connector has been plugged in the vehicle or not, which is an event that is detectable (once the relay is closed) by a change in the 12V voltage carried by the CP wire itself. With a timer involved. This, I am not so sure HAA can do it.

If you want to test the original solution above using the Sonoff SV + the Arduino (retaining the automatic deactivation of the ability to charge after time up) with HAA instead of eWeLink; or the Sonoff SV only with HAA and without the Arduino (hence without the automatic disabling feature) then flash the Sonoff SV with HAA and use this MEPLHAA script:

`{"c":{"io":[[[12,13],2],[[0],6]],"l":13,"b":[[0,5]],"n":"evse"},"a":[{"0":{"r":[[12]]},"1":{"r":[[12,1]]},"b":[[0]]}]}`

The integrated button on the Sonoff SV board (GPIO 0) can be hold over 8 seconds to access Setup Mode, and is also a momentary switch to enable or disable charging on the EVSE physically.
